### DISCUSSION TITLE
Recommended way for processing .env vars outside the framework?

### DISCUSSION THREAD

**Discussion by benr77** [symfony/symfony#47771] [Q&A]
Hello. We are having trouble using the DotEnv component outside of the Symfony framework. See https://github.com/phpro/grumphp/issues/1042 .

We have an application which is building its own container, but the values of the env vars we define are not being resolved correctly.

```
# .env.local
GITLAB_TOKEN=XXXXXX
```

```
# Our config file
api_token: '%env(GITLAB_TOKEN)%'
```

Instead of making the actual value of the GITLAB_TOKEN env var available, I get a strange string which looks like `env_e7937a168f382d27_GITLAB_TOKEN_4e06944bcec1f665r4a82bf0125e3f8d`.

Symfony dependency-injection discovers that we want to use a ENV var:
https://github.com/symfony/dependency-injection/blob/b9c797c9d56afc290d4265854bafd01b4e379240/ParameterBag/EnvPlaceholderParameterBag.php#L72

Yet it fails to resolve that - even though the https://github.com/symfony/dependency-injection/blob/6.1/Compiler/RegisterEnvVarProcessorsPass.php goes over it.

It seems like the container does not provide any container.env_var_processor by default. Meaning the env var doesn't get resolved into it's corresponding value?

Are we missing something? What is the recommended way to do this outside of the Symfony framework?

Thanks

**Comment by gaborjonas** [comment]
I've just spent 2 hours on the same issue, you'll need `$container->compile(true);`
https://github.com/symfony/dependency-injection/blob/7.0/ContainerBuilder.php#L735

**Reply by benr77** [reply]
Thank you for this!
